### DISCUSSION TITLE
The ERB challenge

### DISCUSSION THREAD

**Discussion by sultanmyrza** [rails/rails#53079] [Ideas]
Is anyone else struggling with ERB formatting? When I combine ERB + Tailwind with Stimulus, the formatting still feels off.

I recently watched @dhh's [opening keynote](https://www.youtube.com/watch?v=-cEn_83zRFw) and was super excited. Right after the keynote, I scaffolded a new project and was amazed by all the features included right out of the box—so many gifts! You get a deploy-ready Ruby on Rails project with solid defaults, Docker support, and Kamal 2 for effortless, Heroku-like deployments. I’m genuinely thrilled with everything Rails offers 🎉.

But then I hit a roadblock with ERB formatting. I love ERB—it’s incredibly powerful, and I really want to keep using it. However, the lack of proper formatting tools is a challenge. As a community, should we push to improve the ERB experience? Maybe it’s time we come together to build a `ruby-lsp-erb` addon and finally provide ERB with the formatting solution it deserves.

I’ve tried various tools like `erb-formatter`, `html-beautifier`, and even `prettier`, but none of them seem to format ERB well enough. It feels like a dedicated LSP addon is what we really need for proper support. I understand ERB isn’t widely used in big tech, so we might not see an LSP ERB formatter anytime soon. However, ERB is quite powerful, especially when combined with Tailwind and Stimulus. Just look at this example showcasing ERB's beauty:

```erb
<h1><%= @article.title %></h1>

<p><%= @article.body %></p>

<ul>
  <li><%= link_to "Edit", edit_article_path(@article) %></li>
  <li><%= link_to "Destroy", article_path(@article), data: {
                    turbo_method: :delete,
                    turbo_confirm: "Are you sure?"
                  } %></li>
</ul>

<h2>Comments</h2>
<% @article.comments.each do |comment| %>
  <p>
    <strong>Commenter:</strong>
    <%= comment.commenter %>
  </p>

  <p>
    <strong>Comment:</strong>
    <%= comment.body %>
  </p>
<% end %>

<h2>Add a comment:</h2>
<%= form_with model: [ @article, @article.comments.build ] do |form| %>
  <p>
    <%= form.label :commenter %><br>
    <%= form.text_field :commenter %>
  </p>
  <p>
    <%= form.label :body %><br>
    <%= form.textarea :body %>
  </p>
  <p>
    <%= form.submit %>
  </p>
<% end %>
```

My ideal scenario is to scaffold a Ruby on Rails project, enjoy all the great features that DHH and the Rails team have crafted with love, and have prettier-like formatting for ERB + Stimulus views right out of the box.

As for ERB formatting, should we build something on top of existing tools like `ruby-lsp`? Perhaps an ERB formatter addon for [ruby-lsp](https://shopify.github.io/ruby-lsp/add-ons.html) would be the right approach. What do others think? How are you handling ERB formatting right now? We can’t talk about linting yet, but getting the formatting right would be a solid first step—one step at a time.

By the way, for Stimulus autocomplete, I’ve been using [Stimulus LSP](https://marketplace.visualstudio.com/items?itemName=marcoroth.stimulus-lsp), and it’s been a game-changer. I just wanted to share it for those who might not know about it yet—it definitely deserves more attention! 

If anyone knows of other tools or extensions that could improve the developer experience (DevEx) when working with ERB, I’d love to hear about them. I’m really interested in discussing what others are using or how they’re handling the view part in Rails. Whether it’s formatting, autocomplete, or anything that makes working with ERB smoother—let's share our ideas and tools to improve things together!

**Comment by EricGusmao** [comment]
Hey, that's a great idea! I've been struggling with the same thing as well.

Maybe we should take a look at how these tools are used and implemented on the laravel side (Blade files), for example.

**Comment by sultanmyrza** [comment]
I also posted [this](https://github.com/rails/rails/discussions/53079#discussion-7246272) question in https://shopify.github.io/ruby-lsp/ Slack workspace, and it seems like there is some work going on: https://ruby-dx.slack.com/archives/C050D17S6FP/p1727499186542949.

> We recently just added request delegation support for better [ERB features](https://shopify.github.io/ruby-lsp/#erb-support). One issue is that request delegation has not yet been standardized as part of the LSP spec, so it requires custom code on the editor side, which means it will only work out of the box for VS Code.
Regarding formatting, we want to add diagnostics and formatting for ERB, we just haven't had the time to finish the work

**Reply by sultanmyrza** [reply]
In the meantime, I’ve been using an LLM to format my ERB code. Every time I hit format, I can’t help but imagine a server somewhere spinning up its GPUs, contributing a tiny bit to global warming just to make my code look pretty. It’s like, “Oops, sorry Earth, but this formatting, indentation is really pretty!” 😅

I feel a bit guilty about it, but hey—the LLM formatter is really good, so I guess I’ll keep melting ice caps one neatly aligned block of code at a time. For now, at least! 😔 

Here is how I ask to "format ERB file"
<img width="1608" alt="Screenshot 2024-10-14 at 9 27 58 PM" src="https://github.com/user-attachments/assets/fe9ca029-48de-49ae-9d66-489174d5135c">

Here’s the result: beautifully formatted code. Honestly, it’s better than any ERB formatter I’ve tried so far. The guilty part? I know I’m melting a bit of ice somewhere with every indent!

<img width="1608" alt="Screenshot 2024-10-14 at 9 28 29 PM" src="https://github.com/user-attachments/assets/4a60fabc-88c0-40a0-ae3e-1d35e82e2e72">
